### PR TITLE
throw exception when creating an entry name that already exists in ZipArchive

### LINKED ISSUE #51051: ZipArchive.GetEntry() doesn't work properly after adding 2 entries with the same path, and removing one of them
### Description

After creating 2 entries with the same path ~(which is allowed by the standard, while not recommended)~ in a `ZipArchive` and deleting one of the entries, `ZipArchive.GetEntry()` w/ the path cannot retrieve the remaining entry, despite being already written into the archive, being able to be retrieved if the archive was readed afterwards. 
#### Minimal reproduce:
```csharp
using System;
using System.IO;
using System.IO.Compression;
public class Program
{
    static void Main(string[] args)
    {
        {
            using var stream = File.Create("..\\..\\..\\testZip.zip") ;
            using var archive = new ZipArchive(stream, ZipArchiveMode.Update);
            var text1 = "1234567890987654321";
            var text2 = "0987654321234567890";
            var previousEntry = archive.CreateEntry("test.txt");
            using (var writer = new StreamWriter(previousEntry.Open()))
            {
                writer.Write(text1);
                writer.Flush();
            }
            using (var writer = new StreamWriter(archive.CreateEntry("test.txt").Open()))
            {
                writer.Write(text2);
                writer.Flush();
            }
            previousEntry.Delete();
            var firstResult = archive.GetEntry("test.txt");
            Console.WriteLine("Looking for test.txt the first time...");
            ValidateResult(firstResult);
        }
        {
            using var stream = File.OpenRead("..\\..\\..\\testZip.zip");
            using var archive = new ZipArchive(stream);
            var secondResult = archive.GetEntry("test.txt");
            Console.WriteLine("Looking for test.txt the second time...");
            ValidateResult(secondResult);
        }
    }
    public static void ValidateResult(ZipArchiveEntry searchResult)
    {
        if (searchResult == null)
        {
            Console.WriteLine("Did not find test.txt in the archive.");
        }
        else
        {
            using var reader = new StreamReader(searchResult.Open());
            Console.WriteLine("Found test.txt in the archive. Content: {0}", reader.ReadToEnd());
        }
    }
}
```
Expected output:
```
Looking for test.txt the first time...
Found test.txt in the archive. Content: 0987654321234567890
Looking for test.txt the second time...
Found test.txt in the archive. Content: 0987654321234567890
```
Actual output:
```
Looking for test.txt the first time...
Did not find test.txt in the archive.
Looking for test.txt the second time...
Found test.txt in the archive. Content: 0987654321234567890
```

### Configuration

dotnet version: 5.0.2 (runtime)
OS & version: Windows 20H2 19042.867
architecture: x64

### Other information

Possible where the issue occured:
When adding an entry, `ZipArchive.CreateEntry()` calls `ZipArchive.DoCreateEntry()`, which then calls `ZipArchive.AddEntry()`, which looks like this:
https://github.com/dotnet/runtime/blob/0af5228e8fbf24f4fecfe292c0df732e64ae13c2/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L421-L430
Here, the attempted addition to `_entriesDictionary` is discarded when multiple entries with the same name occur. This, while prevented an exception from being thrown, caused a loss of information of the second entry, although it's still recorded in `_entries`. 
When deleting an entry, `ZipArchiveEntry.Delete()` calls `ZipArchive.RemoveEntry()`, which looks like this:
https://github.com/dotnet/runtime/blob/0af5228e8fbf24f4fecfe292c0df732e64ae13c2/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L263-L279
which removes the record of an entry with such name, regardless of it being present multiple times.
When searching for an entry:
https://github.com/dotnet/runtime/blob/0af5228e8fbf24f4fecfe292c0df732e64ae13c2/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L329-L340
Here, only the attempt to search for the entry in `_entriesDictionary` happens, therefore unable to retrieve the entry while being present in the archive and `_entries`.

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #51051

### CODE REVIEW

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_CreateTests.cs [hunk 1]
```diff
@@ -178,6 +178,22 @@ public static void CreateNormal_VerifyDataDescriptor()
...
+            using var memoryStream = new MemoryStream();
+            // We need an non-seekable stream so the data descriptor bit is turned on when saving
+            var wrappedStream = new WrappedStream(memoryStream);
+
+            // Creation will go through the path that sets the data descriptor bit when the stream is unseekable
+            using (var archive = new ZipArchive(wrappedStream, ZipArchiveMode.Create))
+            {
+                CreateEntry(archive, "A", "xxx");
+                AssertExtensions.ThrowsContains<InvalidOperationException>(() => CreateEntry(archive, "A", "yyy"),
+                    "The entry name 'A' already exists in the archive.");
```
Comment (carlossanlop) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_CreateTests.cs:193]: @danmoseley do we still have CI legs running in different languages? I once received a suggestion to only look for the variable part of the string (in this case, just the duplicate entry name), not the full exception message.

`AssertExtensions.ThrowsContains` is defined like this:

https://github.com/dotnet/runtime/blob/750965da8d5e5099be46a36480a079e7c7643d44/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs#L23-L27

@pedrobsaila I think you just want to search for the entry name, but I suggest the added entries have a larger unique name. 'A' might be too broad. Something like:

```suggestion
                string entryName = "duplicate.txt";
                CreateEntry(archive, entryName, "xxx");
                AssertExtensions.ThrowsContains<InvalidOperationException>(() => CreateEntry(archive, entryName, "yyy"), entryName);
```

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs [hunk 2]
```diff
@@ -92,20 +92,36 @@ public static void EmptyEntryTest(ZipArchiveMode mode)
...
-                e.LastWriteTime = lastWrite;
-                using (Stream s = e.Open()) { }
+                    ZipArchiveEntry e = archive.CreateEntry("empty.txt");
+                    e.LastWriteTime = lastWrite;
+                    using (Stream s = e.Open()) { }
+                }
+                else
+                {
+                    ZipArchiveEntry e = archive.GetEntry("data2.txt");
+                    e.Delete();
```
Comment (carlossanlop) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs:106]: The changes in this unit test do not seem related to the changes. There's no duplication attempt being tested.

Can you please remove the changes to this file? I think the unit test you added in the other file is doing all the verification we need.

File: src/libraries/System.IO.Compression/src/Resources/Strings.resx [hunk 3]
```diff
@@ -308,4 +308,7 @@
   <data name="Zip64EOCDNotWhereExpected" xml:space="preserve">
     <value>Zip 64 End of Central Directory Record not where indicated.</value>
   </data>
+  <data name="EntryNameAlreadyExists" xml:space="preserve">
+    <value>The entry name '{0}' already exists in the archive.</value>
```
Comment (carlossanlop) [inline src/libraries/System.IO.Compression/src/Resources/Strings.resx:312]: Tiny suggestion - how about:
```suggestion
    <value>An entry named '{0}' already exists in the archive.</value>
```

**Review by carlossanlop — COMMENTED** [review]
Thanks for your contribution, @pedrobsaila , and thank you for your patience.

I left some suggestions for you to consider.

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs (hunk 2 above)
Comment (pedrobsaila) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs:106]: This test attempts to add twice the same files: data1.txt, data2.txt, and empty.txt inside the zip archive when ZipArchiveMode.Update, so the test crashes with the introduced InvalidOperationException exception. The test works fine when ZipArchiveMode.Create because ZipArchive class overwrites the existing zip files.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs [hunk 4]
```diff
@@ -421,12 +426,7 @@ internal void AcquireArchiveStream(ZipArchiveEntry entry)
         private void AddEntry(ZipArchiveEntry entry)
         {
             _entries.Add(entry);
-
-            string entryName = entry.FullName;
-            if (!_entriesDictionary.ContainsKey(entryName))
-            {
-                _entriesDictionary.Add(entryName, entry);
-            }
+            _entriesDictionary.Add(entry.FullName, entry);
```
[the remark this answers is not in the document]
Comment (pedrobsaila) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs]: seems fair to me

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs (hunk 2 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs:106]: Ah, I see what you mean. Before your changes, the `AddEntry` call would crash.

Still, my comment applies: the current change isn't testing your new code.

I suggest you remove your if elses, keep the same code as before, _but_ the `AddEntry` invocations inserting a duplicate entry should be wrapped in an `Assert.Throws`. That way, not only we test your new code, but we also keep going with the rest of the test code, and there shouldn't be a duplicate entry accidentally inserted in the end (the last line converting the streams to arrays and comparing them would verify this).

Agree?

Comment (pedrobsaila) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs:106]: I can't remove the if/else, because the InvalidOperationException is thrown only in case of ZipArchiveMode.Update. The case with ZipArchiveMode.Create works fine.

Comment (pedrobsaila) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs]: Sorry for my bad english, maybe I didn't explain well my point of vue. The test to pass needs to be like this :

https://github.com/pedrobsaila/runtime/blob/2887dbee99f2c211a16782adc83563281fbfa757/src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs#L91-L125. So I rollbacked the code and catched the exception but still letting the if/else

**Review by carlossanlop — APPROVED** [review]
Looks great, @pedrobsaila. Thanks for your contribution!

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs [hunk 5]
```diff
@@ -388,6 +388,11 @@ private ZipArchiveEntry DoCreateEntry(string entryName, CompressionLevel? compre
             if (_mode == ZipArchiveMode.Read)
                 throw new NotSupportedException(SR.CreateInReadMode);
 
+            if (_entriesDictionary.ContainsKey(entryName))
+            {
+                throw new InvalidOperationException(string.Format(SR.EntryNameAlreadyExists, entryName));
+            }
```
Comment (jonathanpeppers) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:394]: I'm working on getting various MAUI workloads running on .NET 7. One of our tests hit an issue due to this change.

An MSBuild task inside a NuGet package has the following call:

https://github.com/xamarin/XamarinComponents/blob/895182cc16d88b0e092578d8a6d27bb68d0d459a/Util/Xamarin.Build.Download/source/Xamarin.Build.Download/AndroidAarFixups.cs#L60

Previously the code was creating a new entry then removing the old one afterward. (Code is old, it may indeed be incorrect)

Is the new exception intentional? Is it something that needs to be on a "breaking changes" list somewhere?

Comment (jonathanpeppers) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:394]: I filed an issue; it seems I can reproduce in a smaller example: https://github.com/dotnet/runtime/issues/68734
